ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -937,6 +890,323 @@ chainSyncClient mkPipelineDecision0 tracer cfg
+             -- fragment so they keep in sync.
+             trimmedHeaderStateHistory' =
+               HeaderStateHistory.trim
+                 (AF.length trimmedCandidateFrag)
+                 theirHeaderStateHistory
+
+      | otherwise ->
+        -- No more intersection with the current chain
+        return LostIntersection
+  where
+    ChainDbView {
+        getCurrentChain
+      } = cdbView
+    KnownIntersectionState {
+        theirFrag
+      , theirHeaderStateHistory
+      , ourFrag
+      } = kis
+
+{-------------------------------------------------------------------------------
+  Acquiring a ledger view for RollForward validation
+-------------------------------------------------------------------------------}
+
+data LedgerViewCheck blk =
+    -- | The upstream chain no longer intersects with our current chain because
+    -- our current chain changed in the background.
+    NoLongerIntersects
+    -- | The upstream chain still intersects with our chain, return the latest
+    -- 'KnownIntersectionState' and the 'LedgerView' necessary to validate the
+    -- RollForward header.
+  | LedgerViewAcquired
+      (KnownIntersectionState blk)
+      (Ticked (LedgerView (BlockProtocol blk)))
+
+-- | The peer has sent a RollForward header but their chain is not dense enough
+-- for us to validate the header in the usual way (via forecasting). We'll have
+-- to pre-emptively fetch some of their blocks in order to do so.
+--
+-- This is the state type for the logic requests fetchs and processes the
+-- fetched blocks.
+data LowDensity blk
+  = LowDensity
+      !(KnownIntersectionState blk)
+      -- ^ The 'mostRecentIntersection' via which we obtain the
+      -- initially-unadvanced 'LedgerState' (whose 'LedgerView' forecast did
+      -- not reach the slot of the RollForward header).
+      !(LedgerState blk)
+      -- ^ The advanced ledger state so far.
+      ![RealPoint blk]
+      -- ^ The remaining blocks on the peer's chain to fetch and then apply in
+      -- order to further advance this ledger state.
+      !RequestStatus
+      -- ^ Whether the candidate fragment is certainly set to the next point.
+  deriving (Generic)
+  deriving anyclass (NoThunks)
+
+data RequestStatus = RequestReady | RequestStale
+  deriving (Generic, Eq)
+  deriving anyclass (NoThunks)
+
+-- | Attempt to forecast the necessary ledger view from the
+-- 'mostRecentIntersection'. If that doesn't work, enter the supplied
+-- 'LowDensity' handler.
+checkDensityOrContinue :: forall blk m.
+     (MonadSTM m, LedgerSupportsProtocol blk)
+  => TopLevelConfig blk
+  -> ChainDbView m blk
+  -> SlotNo
+     -- ^ forecast to this slot
+  -> Stateful (STM m) (LowDensity blk) (m (LedgerViewCheck blk))
+     -- ^ how to handle a low density chain
+  -> Stateful (STM m) (KnownIntersectionState blk) (m (LedgerViewCheck blk))
+checkDensityOrContinue cfg cdbView hdrSlot kont = Stateful $ \kis -> do
```

Comment:
I typically go for `k`, it's almost as common as using `f` and `g` for function arguments. Sidenote: "kont" means "butt" in Dutch, but we're grownups, so I don't care.
```suggestion
checkDensityOrContinue cfg cdbView hdrSlot k = Stateful $ \kis -> do
```